### PR TITLE
Fix for issue with upgrading projects generated prior JHipster 3.4.0 #3761

### PR DESCRIPTION
**Pull request by Invizible**
I've added a task which installs jhipster locally right before the app generation, so now when you run 
`yo jhipster:upgrade`
it will:

1. check out on the jhipster_upgrade branch
2. clean up the folder (_cleanUp task)
3. install jhipster with needed versioin
4. and then run _generate task, which will use installed locally jhipster and generate correct app

So now upgrade process is independent of the installed globally version and we can upgrade projects generated before 3.4.0. And it also fixes issues when we have several projects generated with different jhipster versions.

### CODE REVIEW

**Comment by Invizible** [comment]
I've tested fix on two projects 3.0.0 and 3.4.0

**Comment by deepu105** [comment]
This makes the sub gen even slower as there will be 2 npm installs. Check my comment on the issue. I dont think we should do this

File: generators/upgrade/index.js [hunk 1]
```diff
@@ -155,12 +154,24 @@ module.exports = UpgradeGenerator.extend({
                 }.bind(this));
             }.bind(this);
 
+            var installJhipsterLocally = function(version, callback) {
+                this.log('Installing JHipster ' + version + ' locally');
+                shelljs.exec('npm install ' + GENERATOR_JHIPSTER + '@' + version, {silent:true}, function (code, msg, err) {
+                    if (code === 0) this.log(chalk.green('Installed ' + GENERATOR_JHIPSTER + '@' + version));
+                    else this.error('Something went wrong while installing generator! ' + msg + ' ' + err);
```
Comment (jdubois) [inline generators/upgrade/index.js]: "while installing the JHipster generator"

**Comment by jdubois** [comment]
I don't know this part well enough -> @lordlothar99 what do you think about this? I see @deepu105 is against it, and personally I don't mind if the upgrade process doesn't work on pre-3.4.0 versions

**Comment by lordlothar99** [comment]
yeah we briefly talked about this in https://github.com/jhipster/generator-jhipster/issues/3761, I wish I could have come on monday to talk about it... 
 in my opinion this would bring value to the upgrader cause there are plenty of projects generated before 3.4.0, and I know guys who told me they were expecting that ; on the other hand I understand @deepu105 argument about performance --> that's why @Invizible refactored a little the code so it wouldn't waste time for "normal" projects. so imo we should  merge this ! @deepu105 what do u think ?

**Comment by deepu105** [comment]
Im ok as long as it doesn't add too much overhead or install time for all

Thanks & regards,
Deepu

**Comment by lordlothar99** [comment]
Yeah!  :sunglasses:

**Comment by Invizible** [comment]
So, are we good with this?

**Comment by deepu105** [comment]
Ok then if everyone is ok with this then Im ok.

File: generators/upgrade/index.js (hunk 1 above)
Comment (deepu105) [inline generators/upgrade/index.js]: @Invizible can you take care of this?

**Comment by deepu105** [comment]
@Invizible i will merge this as soon as you take care of the comment from @jdubois . its a minor wording change

**Comment by Invizible** [comment]
@deepu105  OK, will update PR

**Comment by Invizible** [comment]
@deepu105 done

**Pull request merged**